Parsing and executing commands

copenhas opened the pull request:
Trying to get together code that could parse user input into command structs so they can be executed.

I have no idea what I'm doing. Very exploratory.

copenhas committed:
basic command parsing and prototype of execution

Commands are defined as a struct with the related data from the
command text. Each struct must implement the Command protocol.

Command parsing is generated via macros similar to Phoenix routes. The
placeholders from the command pattern text are used to populate the
fields on the struct you assign.

Command execute is shown as passing in a context struct and providing
a module of functions to allow commands to manipulate the struct finally
returning it. This means that command execution is side effect free and
should be fully testable. Another gain is that the  module of functions
gives a means to restrict what commands are allowed to do, as well as
being able to replicate the API in another language for game object
scripting (ex: Lua).

matthewphilyaw commented on lib/yggdrasil/commands.ex:
```diff
@@ -0,0 +1,110 @@
+  defstruct exit: nil
+end
+
+defimpl Yggdrasil.Command, for: Yggdrasil.Command.MoveCommand do
```
Awesome, love the use of protocols. Seems like the right place to use them. 

One thing though, I believe that you could actually define this inside the module like this

``` elixir
defmodule Yggdrasil.Command.MoveCommand do
  defstruct exit: nil

  defimpl Yggdrasil.Command, for: Yggdrasil.Command.MoveCommand do
    def execute(player, cmd) do
      Yggdrasil.Player.notify player, "You head towards #{cmd.exit}"
    end
  end
end
```

Generally how I have seen it/done it.

copenhas replied:
Yeah at the time I wasn't sure and didn't want to try. Was too concerned about figuring out how I wanted things to look and work. I'll try to split it out when I have something working.

copenhas commented on test/lib/yggdrasil/command/parser_test.exs:
```diff
@@ -0,0 +1,17 @@
+defmodule ParserTest do
```
@matthewphilyaw things are starting to come together now. Did a bit of clean up and added these tests.

matthewphilyaw commented on test/lib/yggdrasil/command/parser_test.exs:
```diff
@@ -0,0 +1,17 @@
+  end
+
+  test "parse/1 returns :error with reason when parsing is unsuccessful" do
+    assert {:error, _reason} = Parser.parse "IHopeOneDay ThisIsn'tACommand"
```
This brings up on a interesting question, is this invalid as well

``` elixir
{:ok, cmd} = Parser.parse "MoVe EAST"
```

It makes me wonder if we should force lower case for all commands. I only ask because the way we have it currently it would not match, but is MOVE east, wrong?

copenhas replied:
Yeah that's a good point. I could make the `parse/1` function lower case the string before splitting it.

copenhas replied:
Just to follow up, I made the changes in the `ParseRule` to down case the command pattern string. Then I updated the generated `parse/1` function to down case the input string.

matthewphilyaw commented on lib/yggdrasil/command/parser_builder.ex:
```diff
@@ -0,0 +1,64 @@
+
+    quote do
+      def parse(text) do
+        do_parse_command(String.split(text))
```
I wonder based on my comment in the test section if we shouldn't force this to lower case or upper, same for the parse rules. So like parse "GO :exit" vs parse "go :exit" - while I guess it could be argued that if you type GO vs go you intended that, but if the user types go instead of GO they still intended to go... 

Don't know your test case with the mixed case just made me think about this.

matthewphilyaw commented on lib/yggdrasil/command/parser_builder.ex:
```diff
@@ -0,0 +1,64 @@
+
+  defmacro parse(text, command) do
+    quote do
+      @parse_rules ParseRule.create(unquote(text), unquote(command))
```
Would you need to downcase here as well? So that text is forced lower

Thinking about this in the parser

`parse "GO :exit" vs parse "go :exit"`

copenhas replied:
Yep. Would have to do it for the macro and the generated function.

copenhas commented on lib/yggdrasil/player.ex:
```diff
@@ -12,9 +13,8 @@ defmodule Yggdrasil.Player do
-    player_pid = Registry.get_player user_id
-    GenServer.cast(player_pid, {:run_cmd, cmd})
+  def run_cmd(user_id, cmd) do
+    Command.execute cmd, user_id
```
I did this so the calling process would crash if `cmd` doesn't implement the `Command` protocol. I think that's what we want.

matthewphilyaw replied:
Yeah I agree with the crashing however one thing I'm curious about and noted above do we want the command to know about the player? 

This is the player module, it know's how to talk to players, so I kind of would rather see

``` elixir
def run_command(user_id, cmd) do
  notify user_id, Command.execute(cmd)
end
```

what concerns me is the side effects in Command.execute. Like for now it's simple but just kind of worried as this grows into thinks like attack commands and other action commands what does that look like.

Or maybe another approach would be to delegate to a fun.

``` elixir
def execute(cmd, fun) do
  # Think the fun take in a keyword list
  # such that fun has things like [exit: north, message: "you head north"]
  # and the caller can pattern match out the values.
  fun.([message: "You head #{cmd.exit}"]) 
end

# player module
def run_cmd(user_id, cmd) do
  Command.execute(cmd, fn args -> notify(user_id, args[:message]))
end
```

don't know, like I get the intent, just don't like the direct side effect there...

matthewphilyaw replied:
Like I'm not sure commands need to perform actions, but rather be the context for something that know's how to handle the command. I think that is what I'm trying to articulate.

copenhas replied:
Well something will need to do it but I think I see your point. My original intention was that the `user_id` would be able to help us identify the room they are in, their character, anything else the command might need to perform it's logic.

I'm certainly not married to the design. Perhaps another idea is that the room executes all commands and passes in the room struct and character id then it returns the updated room struct? Then when you issue a command the player process would lookup the room and forward the command with who submit it. Only sketchy part is we wouldn't want to crash the room if something was wrong with the command. I suppose the room could spawn a process for the command. Yeah I don't know. More thought will need to be done.

matthewphilyaw replied:
I think what put me off was the player being in the move command, and well we don't have anything else so you put what we have and the more I think through it the protocol would probably be the best way to dispatch for now. So sorry for the initial reaction, that thing I can't put my finger on about what I don't like I think is the lack of a room so it looks odd right now haha.

What I'm hung up on at the moment is that way execute is defined to take a cmd and user_id -- would there ever need to be more? So what I have mocked up below introduces a context object, and the context is created in the player command, and passed into the command.execute function.

What came to mind is the way Plug works and how it's the backbone of phoenix. Plug is centered around a connection and you have a large number of functions that center around taking in a connection mutating and returning and it's very composable.

I wonder if we can't take a cue from this and do something similar which would then make this context object the backbone of a command in a game, then we can write smaller function that compose together.

``` elixir
defmodule Yggdrasil.Command.MoveCommand do
  defstruct exit: nil
  import Game.context

  defimpl Yggdrasil.Command do
    alias Yggdrasil.Message

    def execute(context, %MoveCommand{exit: direction}) do
      context 
      |> assigns(exit: direction)
      |> Room.exit 
      |> Room.enter
    end
  end
end

defmodule Room do
  import Game.Context

  # looks in assigns for room to enter, and enters room
  # could handle spawning it if it's not started.
  # the room would talk to the user.
  def enter(ctx = %Context{}) do
    room_to_enter = ctx.assigns.enter
    # ... 
    done(ctx)
  end

  # exits current room would send message "You head .."
  # operates on context and the current room knows it's exits perhaps
  # and injects the rooms id into the assigns
  def exit(ctx = %Context{}) do
    room = find_current_room(ctx.user_id)
    exit_user(ctx.user_id, room) # runs some game logic to talk to player
    next_room =  find_exit_id(room, ctx.assigns.exit)
    assigns(ctx, enter: next_room)
  end

  # define all the helper functions for a room like find_current_room etc..
end

# player module

def run_cmd(user_id, cmd) do
  # first idea, if it turns out that we only ever pass user_id
  # then maybe we just move the Context.new into the commands
  # and restore this back to user execute Command.execute(cmd, user_id)
  Game.Context.new
  |> add_user(user_id)
  |> Command.execute(cmd)
end
```

copenhas replied:
Just as a heads up you gotta put the `cmd` first because the protocol will do dynamic dispatch off of the first argument. 

Yeah I was trying to suggest something more like that above where maybe the room executes the command and passing in a room struct. But yeah there might need to be a custom thing created so commands have everything available to them. Naively it could be `%{ player: %Character{}, room: %Room{}, actions: [] }` or something. The `:actions` key is for keeping track of everything the command triggered. If we wanted to keep commands side effect free then the functions the command uses could tack on what should happen into the `:actions` key. It pushes the side effects to the `Room` and makes each command I think unit testable.

Dummy example of what a `push button` command could trigger:

``` elixir
def execute(cmd, ctxt) do
    ctxt
    |> spawn_creature(Troll)
    |> notify_player(Message.info("You have sprung a trap!")
    |> set_timeout({:second, 60}, %ResetTrapEvent{ trap: cmd.target }) 
end
```

So the resulting context object would have 3 actions in it for the room to perform.

matthewphilyaw replied:
Ah yeah wasn't paying attention to position got caught up in the pipeline. 

That seems reasonable to me. You build up this context and the room executes it. 

I like the idea of this expanded structure you just roughed out. I wasn't opposed to just strongly to a vanilla room struct but felt limited. I think what you just proposed would be a great start. 

Then actions like spawn_creature would append the action but the room could do the spawn so unit tests are reduced to just checking the context after a command executes. 

So the player process could execute command then pass the context to something like 

ctx = Command.execute cmd, ctx
Room.run ctx

Or something.

copenhas replied:
Cool sounds like we agree on an initial design. I'll try to get something together as a prototype until we can get the room and world stuff developed.
- Sean Copenhaver

> On Jan 27, 2016, at 11:17 PM, Matthew Philyaw notifications@github.com wrote:
> 
> In lib/yggdrasil/player.ex:
> 
> > @@ -12,9 +13,8 @@ defmodule Yggdrasil.Player do
> >      Supervisor.add_player user_id, [owner_pid, push_msg]
> >    end
> > -  def run_cmd(user_id, cmd = %Message{type: :command}) do
> > -    player_pid = Registry.get_player user_id
> > -    GenServer.cast(player_pid, {:run_cmd, cmd})
> > -  def run_cmd(user_id, cmd) do
> > -    Command.execute cmd, user_id
> >   Ah yeah wasn't paying attention to position got caught up in the pipeline.
> 
> That seems reasonable to me. You build up this context and the room executes it.
> 
> I like the idea of this expanded structure you just roughed out. I wasn't opposed to just strongly to a vanilla room struct but felt limited. I think what you just proposed would be a great start.
> 
> Then actions like spawn_creature would append the action but the room could do the spawn so unit tests are reduced to just checking the context after a command executes.
> 
> So the player process could execute command then pass the context to something like
> 
> ctx = Command.execute cmd, ctx
> Room.run ctx
> 
> Or something.
> 
> —
> Reply to this email directly or view it on GitHub.

matthewphilyaw replied:
The only thing that I thought of this morning was that I think the command should be executed in the player process not in the caller.

The player process is the players context and I would imagine just about all commands would need that in keeping with your example

``` elixir
%{
  player: %Character{},
  room: %Room{},
  # ...
}
```

if we execute the command in the player process we know the character and room I think, and not have to look this up. I think that is more natural starting point for the execution of command.

A player process dying should be ok, and if the command failed it could be player related anyway.

copenhas replied:
I was thinking the command would be executed by the room process. It would sort of be the coordinator between various things anyway and it would naturally serialize things like trying to pick up an item before someone else, or attacking something before it leaves. The room could also wrap the execute in a `try...catch` and forward the error back to the player. This is just my initial design idea.

matthewphilyaw replied:
Was referring to this idea you had above

``` elixir
def execute(cmd, ctxt) do
    ctxt
    |> spawn_creature(Troll)
    |> notify_player(Message.info("You have sprung a trap!")
    |> set_timeout({:second, 60}, %ResetTrapEvent{ trap: cmd.target }) 
end
```

I saw that as building a context not doing anything. I may have interpreted that wrong, sorry. 

To me if we did that then the commands are very testable and maybe execute is the wrong word for that function and with that in mind, I was thinking that player process is an ideal starting place. So the following is what I was thinking with your example above applied to move command.

``` elixir
# move command
def execute(cmd, ctx) do
  ctx
  |> move_player(cmd.exit)
  |> notify_player("You head #{cmd.exit}")
end

# player module
def handle_cast({:run_cmd, cmd}, _, state) do
  ctx = Game.Context.new
  |> add_character(state.character)
  |> add_room(state.room)

  ctx = Command.execute(cmd, ctx)

  Room.run ctx
end

#-----------------------------------------
# so after this
ctx = Game.Context.new
|> add_character(state.character)
|> add_room(state.room)

ctx = Command.execute(cmd, ctx)

# I imagine ctx looking like this
%Context{
  player: %Character{name: "that guy"},
  room: %Room{id: 123},
  actions: [
    move_player: {:exit, "north"},
    notify_player: {:message, "You head north"}
  ]
}
```

One maybe execute is the wrong function name for my example, and it should be called build - but once the context is built (shown in the last part of the example above) that is very easy to test and verify no side effects have happened yet.

Room.run (what ever we call it) would be the mechanism that dispatches a context to a room, and I would think could be how player and room interacts, and because the ctx passed in has the room info and player the Room.run can dispatch it and the room knows the player the command belongs to.

The idea comes from Ecto, where you apply several changes to be performed in a changeset validate them and then hand it to the Repo to apply the changes. I'm not sure this is all that different.

matthewphilyaw replied:
That said, I think both agree that the room should handle the actual "execution" of the events and do agree that it would naturally serialize the actions.

All I was suggesting is that commands build the context and we dispatch the context object to the room and as it works through the actions it creates all the side effects etc.

copenhas replied:
So I think the room should "execute" the command because I don't want to worry about a command succeeded but then the room saying "oh you failed because that doesn't exist anymore". Having the room execute the command then immediately handle the actions triggered means we don't have to worry about that. The command can know.
- Sean Copenhaver

> On Jan 28, 2016, at 3:24 PM, Matthew Philyaw notifications@github.com wrote:
> 
> In lib/yggdrasil/player.ex:
> 
> > @@ -12,9 +13,8 @@ defmodule Yggdrasil.Player do
> >      Supervisor.add_player user_id, [owner_pid, push_msg]
> >    end
> > -  def run_cmd(user_id, cmd = %Message{type: :command}) do
> > -    player_pid = Registry.get_player user_id
> > -    GenServer.cast(player_pid, {:run_cmd, cmd})
> > -  def run_cmd(user_id, cmd) do
> > -    Command.execute cmd, user_id
> >   That said, I think both agree that the room should handle the actual "execution" of the events and do agree that it would naturally serialize the actions.
> 
> All I was suggesting is that commands build the context and we dispatch the context object to the room and as it works through the actions it creates all the side effects etc.
> 
> —
> Reply to this email directly or view it on GitHub.

copenhas commented on web/channels/player_channel.ex:
```diff
@@ -34,7 +34,12 @@ defmodule Yggdrasil.PlayerChannel do
   def handle_in("player_cmd", %{ "text" => text }, socket) do
-    Player.run_cmd(socket.assigns.user, Message.command(text))
-    {:noreply, socket}
+    case Parser.parse(text) do
```
With this everything is wired up now. If you type `move east` you'll get the message from the command. Kind of cool to see if working. I guess I need to write some tests to cover what I can and move on to the next thing.

matthewphilyaw commented on web/channels/player_channel.ex:
```diff
@@ -34,7 +34,12 @@ defmodule Yggdrasil.PlayerChannel do
+        Player.run_cmd(socket.assigns.user, cmd)
+        {:noreply, socket}
+      {:error, reason} ->
+        {:reply, {:error, Message.error(reason)},  socket}
```
I don't think we want to `{:reply, {:error, ..}}` here. Client doesn't see the message, which is ok it can be made aware but in particular the reason I'm against this is how the message is manifested on the web socket. 

``` json
/*bogus command */
{"topic":"player:1","ref":"8",
 "payload":{"status":"error","response":{"type":"error","message":"Unknown command.\n"}},
 "event":"phx_reply"}
 /* in particular event type "phx_reply" doesn't seem productive for catching game errors */

/* good command*/
{"topic":"player:1","ref":null,
 "payload":{"type":"info","message":"You head towards east"},
 "event":"event"}
```

While the parse failed, which is an error I think it may be more beneficial to either manifest it back on the 'event' channel and have on place to listen to it, as it's game response (there will be many you can't do that's I'm sure) or maybe we do something like

``` elixir
{:error, reason} ->
  # wrap in a payload if need be just a quick example (can't remember the format of Messages...)
  broadcast! socket, "error", Message.error(reason) 
```

This give a better point to latch onto for errors, because the phx_reply subject has several message on it including a heartbeat, so if we latched on to it we would have to constantly filter out that heart beat and ensure it was "game" error we are displaying.

copenhas replied:
Yeah that's a great idea. I didn't think about what that status might translate to on the client. I was just trying to get everything working together on the server.

matthewphilyaw replied:
I didn't either till I typed in a command wrong. :smiley: and nothing crashed nor did I see anything anywhere until I looked at WS payload in the browser debugger.

matthewphilyaw commented on lib/yggdrasil/command/move.ex:
```diff
@@ -0,0 +1,12 @@
+  alias Yggdrasil.Message
+
+  def execute(cmd, player) do
+    Yggdrasil.Player.notify player, Message.info("You head towards #{cmd.exit}")
```
Do we really want the execute to be aware of the player? Not sure it needs to be... feels out of place, I'll make a note down below where it's called and explain why I think so.

matthewphilyaw commented on lib/yggdrasil/command/move.ex:
```diff
@@ -0,0 +1,12 @@
+  defstruct exit: nil
+end
+
+defimpl Yggdrasil.Command, for: Yggdrasil.Command.MoveCommand do
```
Btw, this can be inside the module I think that is more idiomatic.

``` elixir
defmodule Yggdrasil.Command.MoveCommand do
  defstruct exit: nil

  defimpl Yggdrasil.Command do
    alias Yggdrasil.Message

    def execute(cmd, player) do
      Yggdrasil.Player.notify player, Message.info("You head towards #{cmd.exit}")
    end
  end
end
```

copenhas commented on lib/yggdrasil/player.ex:
```diff
@@ -12,9 +13,16 @@ defmodule Yggdrasil.Player do
+    Enum.each ctxt.actions,
+      fn {:notify, player, message} ->
+        notify(player, message)
+      end
```
I know this looks terrible but the idea is to separate who performs the mutations which this accomplishes. I think once a game world is built and rooms are handling commands we'll move this all out.

copenhas replied:
Hmm... actually a room won't always be handling the commands. If you haven't joined the world yet, or if you are building a character, then the commands need to go somewhere else. I suppose an easy way of handling this is make the game selection and character creation all part of the main web app. Once completed it drops them into the game world so a room would always be handling commands.

matthewphilyaw replied:
Yeah, actually very similar to what I was proposing just instead of executing the command then looping through the context the context would be built first then sent to the room and the room applies it by looping and performs the mutation. 

``` elixir
# probably should be build
def build(cmd, ctx) do
  ctx
  |> move_player(cmd.exit)
  |> notify_player("You head #{cmd.exit}")
end

# player module
# by starting in the process instead of the caller
# we know the player's state and the room it's in I think it makes sense to put it here
def handle_cast({:run_cmd, cmd}, _, state) do
  ctx = Game.Context.new
  |> add_character(state.character)
  |> add_room(state.room)

  # nothing has been done at this point, no side effects.
  # just a pure data structure
  # this step just adds stuff to the context 
  ctx = Command.build(cmd, ctx)

  Room.run ctx
end

# room module

# absolutely agree with you on the need for halting the action chain,
# quick example
def handle_cast({:apply_context, ctx}, _, state} do
  Enum.each ctx.actions, fn action ->
    perform_action(action, state) 
  end
  ...
end

def run(ctx) do
  GenServer.cast(ctx.room, {:apply_context, ctx})
  ...
end

# i realize this take a user_id but not sure we will want that from a room's stand point
def action({:notify, pid, msg}, state) do
  Player.notify pid, msg
  :ok
end

def action({:move, exit}, state) do
  case Enum.any? state.exits, fn ex -> ex == exit end do 
    true -> :ok
    false -> {:error, "#{exit} doesn't exist in this room"}
  end
end
```

I don't know like I'm good with what you are doing as well, mainly was trying to eliminate the commands in the rooms, it seems like they could operate entirely off of context which means we could test the commands in isolation as they just produce a context, and we could test rooms without a command and have fine control over the context during testing to exercise all the actions.

So for now I would just stay on the path you are own for now, as we get there will be more for us to point at and articulate different designs. I think the rooms are the part where this all will shake out.

copenhas replied:
My issue is that the context will basically need to know everything about the room so that commands can do what they want. They'll need to know all the contents in the room, whose in the room, exits for the room. The player process won't have any of that info. For me it makes sense that the room receives the command from the player process. It'll know the full context since the player is in the room.

I know the code above is just an example but it does point out one of the issues I pointed out earlier. You add the move action and a notify action, but then later the move can fail. Should the code just assume you stop at the action that fails? What if you already told the player something, or notified another person in the room? Without extra code to deal with error conditions for each action the command triggered you could have partial commands or misleading messages.

If the context is the full context of the room, then the command itself can determine what it should do. As an example, the exit does not exist, the command can produce whatever message to the player is appropriate. In the case of picking up an item, the command could check that the item exits, check that the player has the strength to pick it up, check if the item has certain properties, then decide what actions should be triggered.

matthewphilyaw replied:
Ah ok, yeah so I don't see why a context could not be validated first right? In the example I provided the intent of a command gets collapsed down to a list of actions which the room process receives.

So then the room could do this

``` elixir
# room module
def run(ctx) do
  # I would map over the ctc.actions, each one either giving {:ok, action}, or {:error, msg}
  # then check to see if there are any errors before performing the actions.
  ctx = validate(ctx)
end
```

However I now understand what you are doing, and good with that as well. The command's context would get the room struct then the room calls execute and each action in the command's pipeline would probably needs to have some way of effectively not produce anything if there are errors. Ectos changesets do that for example with the valid? key. If something is invalid it sets this and the next function that touches the changeset will effectively be bypassed.

So yeah I'm good with that as well. I think my thoughts were more like it's ok if the context is not valid, it's describing the intent and just being built then it gets validated then applied. This way the commands don't have to know anything about the room just how to build the right forms of the actions to apply.

I'm firmly in the mindset of pipelines so forgive me, didn't mean to drag this out. I was thinking of this in several stages, first stage was start the context, hence why I was wanting to start at the player process, then apply the command to the context transforming (the key is transforming) into actions. Then send this to the room which would then validate the actions, and apply it. I was thinking this makes the commands simpler, as I was saying above it takes out the need for a command to know about the rooms context, it just knows about the variables it was given from the user and that it needs to form an action around it. Then when the context gets in the room the validation runs in the context of the room and has all it's state.

Anyway sorry, I'm good with what you are doing what you just said made perfect sense to me and see where you are headed with it.

copenhas commented on lib/yggdrasil/command/move.ex:
```diff
@@ -0,0 +1,13 @@
+    def execute(cmd, room_ctxt) do
+      room_ctxt
+      |> notify_player("You head towards #{cmd.exit}")
+    end
```
Commands would now compose what they want to happen. They would also have the ability to check if the exit exists for example and produce proper messages to tell the player. Maybe have to also add a `:success` key to the context so if one thing fails the whole command could be ignored.... I don't know. Later concerns. We need so much other stuff in place.

matthewphilyaw commented on lib/yggdrasil/player.ex:
```diff
@@ -12,9 +13,16 @@ defmodule Yggdrasil.Player do
+  def run_cmd(user_id, cmd) do
+    # horrible terrible temporary implementation
+    # room would eventually do all of this
+    ctxt = %Yggdrasil.Room.Context{ player: user_id }
```
So this the part I don't like I think the context should start in the player process and the player process dispatch it to the room. The room can still run the command.

We have a way to resolve user_id to a player process which is good, and I think that is where user_id should stop otherwise it's bleeding to far in.

My thoughts are that if we do this from the player process one, we know the pid of the player so for example `player: self()` can be done, no need for a look up here, and I think the player process would be aware of the room it's in. I sort envision the state of the player process to contain the current room pid, and it talks to the room that way and when the player moves it's just given a new pid to talk to. 

This gets rid of having to have any sort of global look up for rooms etc I think.

copenhas replied:
Yeah I should point out the comment above. Also this was originally done here because I was thinking about the player process crashing. I think this whole function will change completely. We just simply have nothing else.

I think at this point I'm just going to try and bump up the test coverage for things I think will stay and ask for this to be merged in.

matthewphilyaw replied:
Ah yeah, totally agree it's going to look awkward until we have more components. 

I think this is a great start and gives us commands so I'm cool with that. Right now I think what you have we can't really represent any better than you have it -- we lack a room etc and it puts a hook in for context which I think is good.

matthewphilyaw commented on lib/yggdrasil/player.ex:
```diff
@@ -12,9 +13,16 @@ defmodule Yggdrasil.Player do
+      end
   end
 
   def notify(user_id, msg = %Message{}) do
```
I think we will need a Player.send or something in addition to this.

This works great coming from the channel because you have the user_id, but I feel like a room would be more concerned with pids. Neither here nor there at the moment, but just something that came to mind. 

I can see a room maintaining a list of player pids that are currently occupying the room and when something happens in the room it finds the player it's related to or players and sends them a message.

copenhas commented:
Test coverage dropped because I don't think you can effectively test macros with coverage. They are compiled in. I'll have to look into testing the macro parts later.

matthewphilyaw commented:
Yeah I don't think there is a good way to test the macro directly, and macro calls probably should not be included in coverage in that respect. I'm surprised it actually impacted coverage at all, guess it depends on what the tool is looking at -- the source or the compiled file...

matthewphilyaw commented on lib/yggdrasil/command/parser_rule.ex:
```diff
@@ -0,0 +1,29 @@
+defmodule Yggdrasil.Command.ParseRule do
```
We might want to test this or least come back to it. It's relatively simple, so probably only need a few cases but helps build confidence in the functions the parser_builder is using.

matthewphilyaw commented:
Hmm, looking more closely at coveralls now, the Command.Parser module doesn't show any lines to cover so it's 100%. That makes sense and perhaps I misunderstood your comment.

I see it dropping around the player related modules and the shifting around there with the context stuff are there more tests cases needed there and parse_rule isn't tested at all.

copenhas commented:
Well the player related modules I think most of the uncovered code are lines we'll be throwing away shortly as rooms get added. I parse rule struct should probably be tested though. I think I can do that tonight unless you want to do some more testing.

matthewphilyaw commented:
Ah yep good point, it's expected I suppose as well that coverage will juggle around anyway as we are working on pieces to a large puzzle.

Yeah I can test the rule struct seem simple enough.

matthewphilyaw commented:
@copenhas Cool. I started the parse_rule testing, when I had a chance. Definitely could use some expanded as we go with some negative tests as well but gets us started. 

Oh speaking of which those names are bit weird, so the file says parser_rule.ex, but the module says ParseRule, I was unsure which you meant, so I left it that way and followed the same scheme in the tests.

copenhas commented:
Ah thanks for catching that. Uh... `parse_rule` sounds better then `parser_rule` I think.

matthewphilyaw commented:
Cool, yeah that was what I figured was intended didn't have a lot of time so just left as it was. I can change if you prefer or if you are actively working in this right now I can leave it to you to minimize the commits.

copenhas commented:
If you are about to PR tests for it, go ahead and rename the file. I think once that's in I will squash and we'll see about merging this.

matthewphilyaw commented:
Ah actually already in this branch, think I just committed on this branch earlier - sorry about that should have branched and did a PR back into this branch.

https://github.com/voluspa/yggdrasil/commit/b07c13a0e7c2b99aea88451de787037cd141da3a

gives us basic testing. Could use some negative test cases as we progress, but at least ensures the happy path I guess, didn't have a lot of time to bang it out.

copenhas commented:
I rebased and squashed it down to a single commit. You might want to read the commit message first as a sanity check, but ready when you are on the merge.

matthewphilyaw commented on lib/yggdrasil/command/move.ex:
```diff
@@ -0,0 +1,13 @@
+defmodule Yggdrasil.Command.MoveCommand do
+  defstruct exit: nil
+
+  defimpl Yggdrasil.Command, for: Yggdrasil.Command.MoveCommand do
```
@copenhas The only thing I see is this, but probably not worth holding this up, you can drop the `, for: ...` portion so it looks like this

``` elixir
defimpl Yggdrasil.Command do
```

since it's in the module.

That said I'm perfectly ok merging it, just saw it.

copenhas commented:
There that should look better.

matthewphilyaw commented:
Cool, merged it.
